### ISSUE TITLE
Getting segmentation violation error while trying to Run Pkgr on RHEL 

### ISSUE THREAD

**Issue by prashantprash17** [metrumresearchgroup/pkgr#336]
[fsaadmin@ip-XX-XX-XX-XX ~]$ pkgr | grep version
pkgr cli version v2.0.1
[$$@xx.xx.xx.xx ~]$ uname -a
Linux $$@xx.xx.xx.xx 3.10.0-1160.15.2.el7.x86_64 #1 SMP Thu Jan 21 16:15:07 EST 2021 x86_64 x86_64 x86_64 GNU/Linux
[$$@xx.xx.xx.xx ~]$
[$$@xx.xx.xx.xx ~]$ cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.9 (Maipo)
[$$@xx.xx.xx.xx ~]$
[$$@xx.xx.xx.xx ~]$
[$$@xx.xx.xx.xx ~]$ pkgr plan
INFO[0000] Installation would launch 4 workers
INFO[0000] R Version 4.0.3
INFO[0000] OS Platform x86_64-pc-linux-gnu
INFO[0000] found installed packages                      count=0
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x8c6b73]

goroutine 7 [running]:
github.com/metrumresearchgroup/pkgr/cran.linuxSupportsBinary(0xc0001517a0)
        /Users/devinp/go/src/github.com/metrumresearchgroup/pkgr/cran/utils.go:88 +0x33
github.com/metrumresearchgroup/pkgr/cran.SupportsBinary(0x0, 0xc0001517a0)
        /Users/devinp/go/src/github.com/metrumresearchgroup/pkgr/cran/utils.go:74 +0x22
github.com/metrumresearchgroup/pkgr/cran.NewRepoDb(0xc000026100, 0x32, 0xc000025465, 0x3, 0x0, 0x0, 0x1, 0x0, 0x0, 0x0, ...)
        /Users/devinp/go/src/github.com/metrumresearchgroup/pkgr/cran/repodb.go:40 +0xfa
github.com/metrumresearchgroup/pkgr/cran.NewPkgDb.func1(0xc000163e40, 0x4, 0x0, 0x3, 0x0, 0xc00005ec60, 0xc000026100, 0x32, 0xc000025465, 0x3, ...)
        /Users/devinp/go/src/github.com/metrumresearchgroup/pkgr/cran/pkg_nexus.go:37 +0xfd
created by github.com/metrumresearchgroup/pkgr/cran.NewPkgDb
        /Users/devinp/go/src/github.com/metrumresearchgroup/pkgr/cran/pkg_nexus.go:36 +0x264
[$$@xx.xx.xx.xx ~]$

**Comment by dpastoor** [comment]
Hi Prashant,

Thank you for the report, Could you please provide the pkgr.yml used when this error occurred so I can attempt to reproduce

Thanks!

**Comment by prashantprash17** [comment]
Thanks Devin, Below is the pkgr.yml :

[XXX@ip-XX-XX-XX-XX ~]$ cat pkgr.yml
Version: 1
Packages:
  - tidyverse
  - SASxport
  - lattice
  - survival
  - reshape2
  - rmarkdown
  - skimr
  - table1
  - shiny
  - MatchIt
  - optmatch
  - RItools
  - SparseM
  - twang

 

Repos:
 - MPN: http://mpn.metworx.com/snapshots/stable/2020-10-24

 

Library: /packages

**Comment by dpastoor** [comment]
Thanks ok - and then the other piece would be what are the contents of:`/etc/os-release`

I have a feeling its erroring trying to detect the operating system, and given the stack trace I'm pretty sure I know what the fix will be, but the contents of that file will help confirm and I can get you a new version to try.

**Comment by semiga** [comment]
Might be able to help out here with a docker file you can test against 

## RHEL7 docker image (error as above)
```
FROM rstudio/r-base:4.0.3-centos7
USER root
RUN mkdir /test
ENV pkgr_version=2.0.1
ADD https://github.com/metrumresearchgroup/pkgr/releases/download/v${pkgr_version}/pkgr_${pkgr_version}_linux_amd64.tar.gz /tmp/pkgr.tar.gz
RUN tar xzf /tmp/pkgr.tar.gz pkgr && \
    mv pkgr /usr/local/bin/pkgr && \
    chmod +x /usr/local/bin/pkgr
RUN echo -e \
"Version: 1\n\
Packages:\n\
- random\n\
Repos:\n\
- CRAN: \"https://cran.rstudio.com (https://cran.rstudio.com/)\"\n\
Library: /test" > pkgr.yml
RUN pkgr install --debug
```

## Ubuntu bionic (working) 
```
FROM rstudio/r-base:4.0.3-bionic
USER root
RUN mkdir /test
ENV pkgr_version=2.0.1
ADD https://github.com/metrumresearchgroup/pkgr/releases/download/v${pkgr_version}/pkgr_${pkgr_version}_linux_amd64.tar.gz /tmp/pkgr.tar.gz
RUN tar xzf /tmp/pkgr.tar.gz pkgr && \
    mv pkgr /usr/local/bin/pkgr && \
    chmod +x /usr/local/bin/pkgr
RUN echo -e \
"Version: 1\n\
Packages:\n\
- random\n\
Repos:\n\
- CRAN: \"https://cran.rstudio.com (https://cran.rstudio.com/)\"\n\
Library: /test" > pkgr.yml
RUN pkgr install --debug
 ```

**Comment by dpastoor** [comment]
perfect - I can run with that - I am quite confident that the reading of that file (or lack of it) is causing an issue where we go to look up that info referencing a pointer thus it explodes.

**Comment by dpastoor** [comment]
```
⋊> ~ docker run -it --rm rstudio/r-base:4.0.3-centos7 /bin/bash                                                                                                                                      12:49:35
[root@a491c053e5dd /]# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

vs at https://github.com/metrumresearchgroup/pkgr/blob/develop/cran/utils.go#L88 we're looking for Version_Codename as a pointer reference, which in this case would be nil and thus throw the error.

**Comment by prashantprash17** [comment]
Hi ,

[xx@xx-xx-xx-xx ~]$ cat /etc/os-release
NAME="Red Hat Enterprise Linux Server"
VERSION="7.9 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.9"
PRETTY_NAME="Red Hat Enterprise Linux Server 7.9 (Maipo)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.9:GA:server"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

 

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.9
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.9"

**Comment by dpastoor** [comment]
Thanks  - very helpful - so I think I need to do this in a two stage approach. This has caused me to dig into the os-release information across distro's and its more varied than I thought. (This is where we figure out whether to use a specific pattern to look up for linux binaries.

For step 1, I can unblock you all quickly. Second, I want to reassess how to detect redhat vs centos and versions within. For example, the VERSION_ID differences of 7.9 vs 7

I should have a binary for you all to try tomorrow that should be good to go.

**Comment by dpastoor** [comment]
Ok got it working in the centos container with a minimal pkgr yaml file

```
packages:
- R6
Library: /tmp/pkgs
Repos:
- CRAN: https://cran.rstudio.com
```

```
[root@03b726ce6d4f pkgr]# ./pkgr plan
INFO[0000] Installation would launch 12 workers
INFO[0000] R Version 4.0.3
INFO[0000] OS Platform x86_64-pc-linux-gnu
INFO[0000] Package Library will be created               path=/tmp/pkgs
INFO[0002] Default package installation type:  source
INFO[0002] 0:17098 (binary:source) packages available in for CRAN from https://cran.rstudio.com
INFO[0002] Package installation cache directory:  /root/.cache/pkgr
INFO[0002] Database cache directory:  /root/.cache/pkgr/r_packagedb_caches
INFO[0002] package installation status                   installed=0 not_from_pkgr=0 outdated=0 total_packages_required=1
INFO[0002] package installation sources                  CRAN=1 tarballs=0
INFO[0002] package installation plan                     to_install=1 to_update=0
INFO[0002] Library path to install packages: /tmp/pkgs
INFO[0002] resolution time 2.0864365s
[root@03b726ce6d4f pkgr]# ./pkgr install
INFO[0000] R Version 4.0.3
INFO[0000] Package Library will be created               path=/tmp/pkgs
INFO[0000] Default package installation type:  source
INFO[0000] 0:17098 (binary:source) packages available in for CRAN from https://cran.rstudio.com
INFO[0000] Package installation cache directory:  /root/.cache/pkgr
INFO[0000] Database cache directory:  /root/.cache/pkgr/r_packagedb_caches
INFO[0000] package installation status                   installed=0 not_from_pkgr=0 outdated=0 total_packages_required=1
INFO[0000] package installation sources                  CRAN=1 tarballs=0
INFO[0000] package installation plan                     to_install=1 to_update=0
INFO[0000] Library path to install packages: /tmp/pkgs
INFO[0000] resolution time 162.0782ms
INFO[0000] downloading required packages within directory   dir=/root/.cache/pkgr
INFO[0000] downloading package                           package=R6
INFO[0000] all packages downloaded                       duration=318.3841ms
INFO[0000] starting initial install
INFO[0002] Successfully Installed.                       package=R6 remaining=0 repo=CRAN version=2.5.0
INFO[0002] starting individual tarball install
INFO[0002] total package install time                    duration=2.1452382s
INFO[0002] duration:2.1452564s
```

more details incoming

**Comment by dpastoor** [comment]
hey guys - mind giving a whirl the pkg at: https://github.com/metrumresearchgroup/pkgr/releases/tag/v3.0.0-alpha.1

hopefully the problem is fixed in #340

**Comment by semiga** [comment]
Fixed for me!

I tested it across three pkgr.yml files installing [recommended R packages](https://docs.rstudio.com/ide/server-pro/r-package-dependencies.html#r-package-dependencies) for RStudio Server.
Thanks for the quick fix 👍
